### ISSUE TITLE
Rails autoloading issues when used with "async" gem

### ISSUE THREAD

**Issue by morgoth** [rails/rails#44751]
I integrated [async](https://github.com/socketry/async) into the Rails app and I noticed issues with using this gem and Rails autoloder (powered by zeitwerk).

I have the sample app that presents the bug - the setup is visible in this commit https://github.com/morgoth/async-autoload/commit/ed3a76c96cdc46903761226b30db6175f8bf680e

So the problem is that when autoloading some class inside the async tasks, the class is properly loaded only once and all other invocations are failed.

As simple as:
```ruby
Async do |task|
  3.times do |i|
    task.async do
      MyModel.run_something # Problems with autloading
    rescue => e
      puts e
    end
  end
end.wait
```

I tried to produce a simple Zeitwerk script to present a failure, but I couldn't so somehow I can reproduce it in the Rails app only.

It can be workaround by manually requiring the file that should be autoloaded or by setting `config.eager_load = true` in the Rails app.

I guess the "async" gem is the simplest way to parallelize something, but using it within the Rails app causes such problems, so it might not be that easy to popularize it in the Rails community.

My env:
ruby 3.1.1p18
async-2.0.1
rails-7.0.2.3
zeitwerk-2.5.4

This was originally reported on the zeitwerk repo https://github.com/fxn/zeitwerk/issues/214 but we came to conclusion that it has to be some Rails issue as we cannot reproduce this bug with pure zeitwerk.

FYI @fxn @ioquatix @bruno-

**Comment by fxn** [comment]
If you throw

```ruby
require "async"

unless Object.autoload?(:MyModel)
  autoload :MyModel, "#{__dir__}/app/models/my_model.rb"
end

Async do |task|
  10.times do |i|
    task.async do
      MyModel.run_something
    rescue => e
      p $!
    end
  end
end.wait
```

in the root directory of that application, we see that:

* Executed with `ruby`, it consistently succeeds.
* Executed with `bin/rails runner`, it consistently fails (raises some `NameError`s).

That is essentially what Zeitwerk does, and indeed we can check this:

```ruby
require "async"
require "zeitwerk"

loader = Zeitwerk::Loader.new
loader.push_dir("app/models")
loader.setup

Async do |task|
  10.times do |i|
    task.async do
      MyModel.run_something
    rescue => e
      p $!
    end
  end
end.wait
```

and it never fails too.

So, the hypothesis is that something in Rails is causing that.

**Comment by matthewd** [comment]
Weird :confused:

> `unless Object.autoload?(:MyModel)`

Lazy question, but maybe you've already checked: what happens if you override/replace the Rails-set autoload with that explicit definition?

(@fxn it has to be Rails configuring Zeitwerk to do something surprising, right? Apart from that setup, I don't think any Rails-proper code gets hit while running something this simple these days?)

**Comment by fxn** [comment]
> Lazy question, but maybe you've already checked: what happens if you override/replace the Rails-set autoload with that explicit definition?

Yeah, I checked. That was only added so that the same script could be run both by `ruby` and `runner`. But if you override, it works too.

> it has to be Rails configuring Zeitwerk to do something surprising, right? Apart from that setup, I don't think any Rails-proper code gets hit while running something this simple these days?)

This is all so simple, that I cannot imagine what might be happening. Need to dig into it.

**Comment by ioquatix** [comment]
I don't think Async does anything odd, it's all fairly standard Ruby code until you do something non-blocking. If I was to suggest a potential issue, it might be that you are locking a mutex but this would only matter in an environment with actual contention because the uncontested mutex should not cause any context switching.

**Comment by fxn** [comment]
I have created a dummy `c.rb` file in the root of the sample project:

```ruby
class C
  def self.run_something
    p self
  end
end
```

and modified the reproduction script to use `C`:

```ruby
require "async"

autoload :C, "#{__dir__}/c.rb"

Async do |task|
  10.times do |i|
    task.async do
      C.run_something
    rescue => e
      p $!
    end
  end
end.wait
```

So, `C` is not loaded by Rails autoloaders.

That script reproduces the pattern: It works with `ruby`, fails with `runner`.

If you know what happens behind the scenes, that script does not bring a lot more information because all Zeitwerk does is defining autoloads on your behalf. But since the issue is weird, every assumption matters.

**Comment by fxn** [comment]
I cannot reproduce in Rails 6.1. Both `classic` and `zeitwerk` mode run the script successfully.

My hypothesis is that something in Rails 7.0 is the culprit. Either in Rails itself, or in some dependency.

**Comment by fxn** [comment]
I have it!

This is produced just by loading the `debug` gem.

This script reproduces without Rails and without Zeitwerk:

```ruby
require 'tempfile'
require 'async'
require 'debug'

Tempfile.create(['foo', '.rb']) do |file|
  file.write(<<~RUBY)
    class C
      def self.call_me
      end
    end
  RUBY
  file.close

  autoload :C, file.path

  Async do |task|
    5.times do |i|
      task.async do
        p C.call_me
        sleep rand
      end
    end
  end.wait
end
```

I think we can close here then. Need to run now, but I guess this has to be reported to `debug`.

**Comment by morgoth** [comment]
Thank you @fxn I opened new issue https://github.com/ruby/debug/issues/580
